### ISSUE TITLE
Have Axe Purchased And Equipped But Not In Inventory And Still In Item Store

### ISSUE THREAD

**Issue by taylorbfoster** [HabitRPG/habitica#2718]
I have purchased the axe for my level 5 character, it is equipped, providing me a strength bonus (both equipment and class-based) and I can see it on my avatar but I cannot see it in my inventory. Additionally the only weapon the item store is currently selling is still the axe, though given I already purchased the axe it should be selling the morningstar. 

Other equipment seems to be functioning properly as I just purchased a leather armor and it is showing up in my inventory and has been replaced in the item shop by the chain mail.

I haven't tried switching it out for another weapon in my inventory in case it is unable to be switched back.

The attach image includes three screenshots taken 5 minutes ago (11:40am AEDT) showing the details of the bugs:
![taylorf_axebug](https://f.cloud.github.com/assets/6570697/2061805/1b83cd66-8c6b-11e3-8996-8f099f599d37.png)

**Comment by deilann** [comment]
Strange bug. What happens if you try to purchase the axe in the store? (You can use fix character values in settings to give yourself the 30 GP to try it out.)

**Comment by taylorbfoster** [comment]
Good idea. After purchasing it (for the second time), the axe is still the only weapon available in the item shop but upon going to my inventory it has appeared there. Returning to the item shop the morningstar has now replaced the axe. Odd.

I'd be tempted to write it off as a failure of syncing but before when I had but didn't have the axe I had left the item shop and returned multiple times with the axe still there (and checked via my Android app) so it seems the act of purchasing did kick it in to gear.

This was all done on PC via Firefox.

**Comment by deilann** [comment]
I'm guessing it was just your PC not syncing up with the server completely. Wonder if this is related to all the 503s.

**Comment by taylorbfoster** [comment]
In that case my PC at home, my Android mobile (fair enough on the same wifi network) AND my PC at work all failed at syncing properly, which suggests to me the syncing issue is either a problem with my account or the server itself. Haven't noticed any 503s!

**Comment by deilann** [comment]
Nah, the server would have been updated with the fact that you have
purchased the axe when it was purchased. So it was a one-time thing, as far
as I can tell. Kind of like it not sticking when you delete a To-Do or move
a Daily.

**Comment by taylorbfoster** [comment]
Have experienced the not-sticking of Dailies, just nothing along the lines of where it seems to have synced in certain areas but not others and still not having synced after a few hours. Will write it off as a one-time syncing issue!

**Comment by deilann** [comment]
yeah, that's what interests me about it. can you
reopen this? i'm not 100% sure i'm right, and i'd like it looked at.

**Comment by taylorbfoster** [comment]
Sure.

**Comment by colegleason** [comment]
I wonder if somehow the purchase op didn't complete but the equip did?

On Sun, Feb 2, 2014 at 9:30 PM, taylorbfoster notifications@github.comwrote:

> Sure.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/issues/2718#issuecomment-33922689
> .

## 

Cole Gleason
Student, University of Illinois at Urbana-Champaign
Computer Science, Class of 2015
cg@colegleason.com | colegleason.com

**Comment by taylorbfoster** [comment]
Sounds plausible, if the purchase op is responsible for both the item shop update AND placing of the item in inventory, both of which seemed to have failed. Wonder at what stage the equip op is called? Assume it would be post-placing of item in to inventory but if it's all assumed to happen simultaneously...

Edit: Look at me, pretending I know how to talk shop...

**Comment by deilann** [comment]
iirc, the purchase op updates the user value that defines both what's in
the store and in your inventory?

**Comment by colegleason** [comment]
Buy op handles both: https://github.com/HabitRPG/habitrpg-shared/blob/develop/script/index.coffee#L611

So that theory seems bust.

**Comment by deilann** [comment]
Although it does go in this order:

```
else
          user.items.gear.equipped[item.type] = item.key
          user.items.gear.owned[item.key] = true
```

So equipping does happen before ownership.

...okay, apparently markdown hates me.

**Comment by colegleason** [comment]
Right, but how the hell does that line not get executed AND the user still
gets saved.

I'd understand if the process died, but then the user would not save.

**Comment by deilann** [comment]
Double requests sometimes leave me with losing whatever happened with the double request. Maybe that was the response here, but there was no dequip. So it fully executed, but then reverted?

**Comment by deilann** [comment]
Basically thinking like when I cast a spell, it doesn't stick, but I still
lose the MP.

**Comment by colegleason** [comment]
Really? That seems odd.

On Sun, Feb 2, 2014 at 11:01 PM, deilann notifications@github.com wrote:

> Basically thinking like when I cast a spell, it doesn't stick, but I still
> lose the MP.
> 
> On Sun, Feb 2, 2014 at 8:37 PM, Cole Gleason <notifications@github.com
> 
> > wrote:
> > 
> > Right, but how the hell does that line not get executed AND the user
> > still
> > gets saved.
> > 
> > I'd understand if the process died, but then the user would not save.
> > 
> > On Feb 2, 2014 10:34 PM, "deilann" notifications@github.com wrote:
> > 
> > > Although it does go in this order:
> > > 
> > > ```
> > > else
> > > user.items.gear.equipped[item.type] = item.key
> > > user.items.gear.owned[item.key] = true
> > > ```
> > > 
> > > So equipping does happen before ownership.
> > > 
> > > On Sun, Feb 2, 2014 at 8:32 PM, Cole Gleason <notifications@github.com
> > > 
> > > > wrote:
> > > > 
> > > > Purchase op handles both:
> > 
> > https://github.com/HabitRPG/habitrpg-shared/blob/develop/script/index.coffee#L611
> > 
> > > > So that theory seems bust.
> > > > 
> > > > ## 
> > > > 
> > > > Reply to this email directly or view it on GitHub<
> > > > https://github.com/HabitRPG/habitrpg/issues/2718#issuecomment-33924224
> > > > 
> > > > .
> > > 
> > > ## 
> > > 
> > > Reply to this email directly or view it on GitHub<
> > > https://github.com/HabitRPG/habitrpg/issues/2718#issuecomment-33924276>
> > > 
> > > .
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub<
> > https://github.com/HabitRPG/habitrpg/issues/2718#issuecomment-33924337>
> > 
> > .
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/issues/2718#issuecomment-33924959
> .

## 

## Cole Gleason

Student, University of Illinois at Urbana-Champaign
Email: cg@colegleason.com
Website: colegleason.com

**Comment by deilann** [comment]
Yeah. A couple of times when I've cast a spell, I've refreshed and lost all
of the benefits of it, but my MP stayed the same, as if I'd actually cast
the spell.

**Comment by taylorbfoster** [comment]
I've only been around for a few days so haven't experienced anything else like that, will leave you to the discussion and let you know if anything else similar occurs!

**Comment by pipcorona** [comment]
Mine did this today too. I bought the shade sword. It equipped it and showed it, but it didn't show in my inventory. I looked in the store and it was there and the gold was back to where it was before the purchase. I refreshed and it was still like that, so I just bought it again and then it worked. The site was having some slowing issues at the time so I assumed it was just a drop in communication somewhere.
